### PR TITLE
Adds kivy namespaces.

### PR DESCRIPTION
**Pull request by matham**
Still need to update docs to give simpler examples.

### CODE REVIEW

File: kivy/uix/behaviors/knspace.py [hunk 1]
```diff
@@ -0,0 +1,433 @@
+'''Provides namespace functionality for Kivy objects. It allows kivy objects
```
Comment (tito) [inline kivy/uix/behaviors/knspace.py]: If you want the documentation generation works well, you need a title:

```
Kivy namespaces
==============

Provides namespace functionality for Kivy objects...
```

**Comment by tito** [comment]
In your first example, what if i open 2 popup, there would be 2 widget for knspace.keyboard, or only the last one would work? Or does it crash?

**Comment by matham** [comment]
If you create two widgets with the same name, but you have only one namespace, then the last one will be the one stored in the namespace. So when you do `knspace.my_widget`, it'll point to the last. There shouldn't be any crashes. If you need to have multiple widgets with the same name you should fork, as in the second example.

File: kivy/uix/behaviors/knspace.py [hunk 2]
```diff
@@ -0,0 +1,433 @@
...
+
+        self._name = value
+        if value:
+            if knspace:
+                setattr(knspace, value, self)
+            else:
+                raise ValueError('Object has name "{}", but no namespace'.
+                                 format(value))
+
+    name = AliasProperty(_get_name, _set_name, bind=('_name', ), cache=False)
```
Comment (dessant) [inline kivy/uix/behaviors/knspace.py]: Wouldn't this conflict with `Screen.name` when used with Screen? How about `knsname` or `kns_name`?

**Comment by matham** [comment]
Docs has been updated to be easy to understand and go from simple to more advanced. This should clarify things. Please read that to understand what the pr is doing.

Also super comprehensive tests have been added and bugs found have been fixed.

**Comment by matham** [comment]
I'm open to changing the name to `kname`, `knsname`, or `kns_name` or whatever is appropriate (I think one of the first two probably).

**Comment by dessant** [comment]
+1 for `knsname`.

**Comment by matham** [comment]
Renamed `name` to `knsname`.

File: kivy/uix/behaviors/knspace.py [hunk 3]
```diff
@@ -0,0 +1,573 @@
...
+
+A motivating example is the example from above:
+
+.. code-block:: kv
+
+    <MyComplexWidget@KNSpaceBehavior+Widget>:
+        Label:
+            text: root.knspace.pretty.text if root.knspace.pretty else ''
+
+    <MyPrettyWidget@KNSpaceBehavior+TextInput>
```
Comment (dessant) [inline kivy/uix/behaviors/knspace.py]: missing `:` tralalalala

File: kivy/uix/behaviors/knspace.py [hunk 4]
```diff
@@ -0,0 +1,573 @@
...
+Then, we do:
+
+.. code-block:: python
+
+    widget = MyComplexWidget()
+    new_knspace = KNSpace()
+    widget.knspace = new_knspace
+
+The rule is that if no knspace has been assigned to a widget, it looks for a
+namespace in it's parent and parent's parent and so on until it find one to
```
Comment (dessant) [inline kivy/uix/behaviors/knspace.py]: its

File: kivy/uix/behaviors/knspace.py [hunk 5]
```diff
@@ -0,0 +1,573 @@
...
+use. If none are found, it uses the default :attr:`knspace`.
+
+When `MyComplexWidget` is created, it still used the default namespace.
+However, when we assigned the root widget it's new_namespace, all its children
```
Comment (dessant) [inline kivy/uix/behaviors/knspace.py]: its

**Comment by dessant** [comment]
I like the api and the docs are well organized and clear. fatty +1

File: kivy/uix/behaviors/knspace.py [hunk 6]
```diff
@@ -0,0 +1,573 @@
...
+    '''The name given to this instance. If named, the name will be added to the
+    associated :attr:`knspace` namespace, which will then point to the
+    `proxy_ref` of this instance.
+
+    When named, one can access this object by e.g. self.knspace.name, where
+    `name` is the given name of this instance. See :attr:`knspace` and the
+    module description for more details.
+    '''
+
+knspace = KNSpace()
```
Comment (tito) [inline kivy/uix/behaviors/knspace.py]: You missed the documentation for that:

    #: default instance of :class:`KNSpace`

**Comment by tito** [comment]
+1 for merging

File: kivy/uix/behaviors/knspace.py (hunk 6 above)
Comment (matham) [inline kivy/uix/behaviors/knspace.py]: This one is documented at the top where we declare the variable: https://github.com/kivy/kivy/pull/3757/files#diff-6e2b6bdb54f0737d905c1f84147e4bf0R269.

**Comment by tito** [comment]
\o/

**Comment by akshayaurora** [comment]
@matham nice work !

**Pull request merged**